zerver/actions/message_send.py

```diff
@@ -210,11 +214,25 @@ def get_recipient_info(
         # of this function for different message types.
         assert stream_topic is not None
 
+        possibly_topic_wildcard_mentioned_user_ids: AbstractSet[int] = set()
+        if possible_topic_wildcard_mention:
+            # Anyone who sent a message to the topic or reacted to a message on the topic
+            # is considered a possible user to be notified for topic mention.
+            messages = messages_for_topic(recipient.id, stream_topic.topic_name)
+            possibly_topic_wildcard_mentioned_user_ids = (
+                {message.sender_id for message in messages}
+            ) | set(
+                Reaction.objects.filter(message__in=messages).values_list(
+                    "user_profile_id", flat=True
+                )
+            )
```

Comment:
> This will make 2 database queries, one to fetch the sender IDs...

`message.sender_id` won't cost an extra db query as the `Message` table already has `sender_id` available.

Ref: https://docs.djangoproject.com/en/4.2/ref/models/fields/#database-representation
> Behind the scenes, Django appends "_id" to the field name to create its database column name.
---
Also, since this query is now getting used in a couple of places ([#20977 > zerver/lib/actions.py](https://github.com/zulip/zulip/commit/1a46df40c33104b685c154a99a57d7bd8dc2c6c0#diff-9a1fd3f8a1bd0ccd686ac2d96d265d3d5dce323f6fc191adac2cb88500af6e1cR6456)),

I'm planning to do a refactoring in a separate commit to have this query inside a lib function, say, `sent_or_reacted_to_messages_user_ids` and use it everywhere.

Reply:
The reason for that being databases do a lot of work to optimise how the result of a query set is calculated. So telling the database to do the OR would most probably be faster than doing it in memory in Python.

Reply:
This is the updated query:
```python
def participants_for_topic(recipient_id: int, topic_name: str) -> Set[int]:
    """
    Users who either sent or reacted to the messages in the topic.
    The function is expensive for large numbers of messages in the topic.
    """
    messages = Message.objects.filter(recipient_id=recipient_id, subject__iexact=topic_name)
    participants = set(
        UserProfile.objects.filter(
            Q(id__in=Subquery(messages.values("sender_id")))
            | Q(
                id__in=Subquery(
                    Reaction.objects.filter(message__in=messages).values("user_profile_id")
                )
            )
        ).values_list("id", flat=True)
    )
    return participants
```

Reply:
Yeah that looks great! 

The further thing we can do is log the actual database query and then do an `EXPLAIN ANALYZE` on that query to make sure it's using good indexes; but I expect it should be.